proposals/yield-in-try-catch.md

```diff
@@ -0,0 +1,500 @@
+    }
+}
+```
+
+This behavior is important to understand when considering the code generation for `try / catch` blocks.
+
+### Code generation of yield inside try with catch in iterators
+
+The code generation for iterators that have `yield` inside a `try / catch` will preserve the `catch` semantics in the `Dispose` method. That is _only_ the contents of the `finally` blocks will execute even in the case exceptions come into play. For example:
+
+```csharp
+var e = M1(true).GetEnumerator();
+e.MoveNext();
+e.Dispose();
+
+IEnumerable<int> M1(bool b)
+{
+    try
+    {
+        try
+        {
+            try
+            {
+                yield return 13;
+            }
+            finally
+            {
+                Console.WriteLine("inner finally");
+                if (b)
+                {
+                    throw new Exception();
+                }
+            }
+        }
+        catch
+        {
+            Console.WriteLine("catch");
+        }
+
+        Console.WriteLine("after catch");
+    }
+    finally
+    {
+        Console.WriteLine("outer finally");
+    }
+}
+```
+
+This program will output:
+
+```cmd
+inner finally
+outer finally
+Unhandled exception. System.Exception: Exception of type 'System.Exception' was thrown
+```
+
+The `"after catch"` is not printed beacuse only the `finally` structure is mirrored in the `Dispose` method. The `catch`, like all other statements between `finally` is not included. Statements in between the `catch` and `finally` are not executed in `Dispose`. This may seem odd at first glance but is leaning into the specified behavior for iterator `Dispose`.
+
+### Dispose and finally in async iterators
+
+The `DiposeAsync` behavior for async iterators mirrors that of traditional iterators in that it executes the `finally` blocks at the the point the state machine is suspended. However it does this by setting the `state` variable to a value that represents disposing and then calls `MoveNextAsync`. The implementation of `MoveNextAsync` will then execute only the `finally` blocks for the suspend point. The `DisposeAsync` method does not have a mirror copy of the `finally` blocks.
+
+### Code generation of yield inside try / catch in async iterators
+
+The code generation of async iterators will change such that `catch` blocks do not observably execute during `DisposeAsync`. To achieve this all `catch` blocks visible from a `yield` will rethrow exceptions if the state machine is in a disposing state. For example consider the following code:
+
+```csharp
+var e = M(true).GetEnumerator();
+await e.MoveNext();
+await e.DisposeAsync();
+
+async IAsyncEnumerable<int> M(bool b)
+{
+    try
+    {
+        try
+        {
+            yield return 13;
+            await Task.Yield();
+        }
+        finally
+        {
+            Console.WriteLine("inner finally");
+            if (b)
+            {
+                throw new Exception();
+            }
+        }
+    }
+    catch
```

Comment:
What if the catch has a `when` clause with side-effects? Would the side-effects be executed?